pkg/capi2mapi/aws.go

```diff
@@ -0,0 +1,432 @@
+func FromMachineAndAWSMachineTemplateAndAWSCluster(m *capiv1.Machine, mts *capav1.AWSMachineTemplate, ac *capav1.AWSCluster) MachineAndAWSMachineTemplateAndAWSCluster {
+	return MachineAndAWSMachineTemplateAndAWSCluster{Machine: m, Template: mts, AWSCluster: ac}
+}
+
+// FromMachineSetAndAWSMachineTemplateAndAWSCluster wraps a CAPI MachineSet and CAPA AWSMachineTemplate and CAPA AWSCluster into a capi2mapi MachineSetAndAWSMachineTemplateAndAWSCluster.
+func FromMachineSetAndAWSMachineTemplateAndAWSCluster(ms *capiv1.MachineSet, mts *capav1.AWSMachineTemplate, ac *capav1.AWSCluster) MachineSetAndAWSMachineTemplateAndAWSCluster {
+	return MachineSetAndAWSMachineTemplateAndAWSCluster{MachineSet: ms, Template: mts, AWSCluster: ac}
+}
+
+// ToProviderSpec converts a capi2mapi MachineAndAWSMachineTemplateAndAWSCluster into a MAPI AWSMachineProviderConfig.
+//
+//nolint:funlen
+func (m MachineAndAWSMachineTemplateAndAWSCluster) ToProviderSpec() (*mapiv1.AWSMachineProviderConfig, []string, error) {
+	if m.Machine == nil || m.Template == nil || m.AWSCluster == nil {
+		return nil, nil, errCAPIMachineAWSMachineTemplateAWSClusterCannotBeNil
+	}
+
+	var (
+		warnings []string
+		errors   []error
+	)
+
+	mapaProviderConfig := mapiv1.AWSMachineProviderConfig{}
+
+	mapaTenancy, err := convertAWSTenancyToMAPI(m.Template.Spec.Template.Spec.Tenancy)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	// The use of ARN and Filters to reference AMIs was present
+	// in CAPA but has been deprecated and then removed
+	// ref: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3257
+	mapaProviderConfig.AMI.ID = m.Template.Spec.Template.Spec.AMI.ID
+
+	mapaProviderConfig.TypeMeta = metav1.TypeMeta{
+		Kind: "AWSMachineProviderConfig",
+		// In the machineSets both "awsproviderconfig.openshift.io/v1beta1" and "machine.openshift.io/v1beta1" can be found.
+		// Here we always settle on one of the two.
+		APIVersion: "awsproviderconfig.openshift.io/v1beta1",
+	}
+	mapaProviderConfig.InstanceType = m.Template.Spec.Template.Spec.InstanceType
+
+	mapiAWSTags := convertAWSTagsToMAPI(m.Template.Spec.Template.Spec.AdditionalTags)
+
+	mapaProviderConfig.Tags = mapiAWSTags
+	mapaProviderConfig.IAMInstanceProfile = &mapiv1.AWSResourceReference{
+		ID: &m.Template.Spec.Template.Spec.IAMInstanceProfile,
+	}
+	mapaProviderConfig.KeyName = m.Template.Spec.Template.Spec.SSHKeyName
+	mapaProviderConfig.PublicIP = m.Template.Spec.Template.Spec.PublicIP
+	mapaProviderConfig.Placement = mapiv1.Placement{
+		AvailabilityZone: ptr.Deref(m.Machine.Spec.FailureDomain, ""),
+		Tenancy:          mapaTenancy,
+		Region:           m.AWSCluster.Spec.Region,
+	}
+
+	mapiAWSSecurityGroups := convertAWSSecurityGroupstoMAPI(m.Template.Spec.Template.Spec.AdditionalSecurityGroups)
+
+	mapaProviderConfig.SecurityGroups = mapiAWSSecurityGroups
```

Comment:
I would probably cuddle these two together since they are related. Do we even need to declare them on separate lines?